### PR TITLE
K8s: Expose datasources as api-servers

### PR DESCRIPTION
**Pull request by ryantxu**
This expands on the work from https://github.com/grafana/google-sheets-datasource/pull/196, but will work for any datasource.

http://localhost:3000/apis/testdata.datasource.grafana.app/v0alpha1/namespaces/default/connections/PD8C576611E62080A/resource

http://localhost:3000/apis/testdata.datasource.grafana.app/v0alpha1/namespaces/default/connections/PD8C576611E62080A/resource/scenarios

http://localhost:3000/apis/testdata.datasource.grafana.app/v0alpha1/namespaces/default/connections/PD8C576611E62080A/health

http://localhost:3000/apis/testdata.datasource.grafana.app/v0alpha1/namespaces/default/connections/PD8C576611E62080A/query?scenarioId=random_walk_table

### CODE REVIEW

**Review by marefr — COMMENTED** [review]
Had a quick look. Looks pretty interesting :+1: 

The config vs instance looks a bit weird I admit, but understand why because we currently don't have any alternative. Earlier you've discussed that list operations might be tricky to support and search would be the solution - is that still the idea?

I guess you could also have a "datasource read only resource" being updated by an operator whenever a datasource instance changes.

File: pkg/registry/apis/datasource/register.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+	info, err := request.NamespaceInfoFrom(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+
+	vals, err := b.dsService.GetDataSourcesByType(ctx, &datasources.GetDataSourcesByTypeQuery{
+		OrgID: info.OrgID,
+		Type:  b.plugin.ID,
+	})
+	// HACK!!! See https://github.com/grafana/grafana/issues/76154
```
Comment (andresmgot) [inline pkg/registry/apis/datasource/register.go:222]: FTR, that issue got solved so you could probably remove this block

File: pkg/registry/apis/datasource/sub_resource.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		http.MethodPost,
+		http.MethodPut,
+		http.MethodPatch,
+		http.MethodDelete,
+		http.MethodOptions,
+	}
+}
+
+func (r *subResourceREST) NewConnectOptions() (runtime.Object, bool, string) {
+	return nil, true, ""
```
Comment (ryantxu) [inline pkg/registry/apis/datasource/sub_resource.go:45]: @andresmgot -- long subpaths should work now.  The second true parameter appears to register everything after the slash also.

Welcome to the mysterious world of kubernetes options!

File: pkg/registry/apis/datasource/sub_health.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+		if len(healthResponse.JSONDetails) > 0 {
+			err = json.Unmarshal(healthResponse.JSONDetails, &rsp.Details)
+			if err != nil {
+				responder.Error(err)
+				return
+			}
+		}
+
+		responder.Object(200, rsp) // TODO... 200 vs ????
```
Comment (ryantxu) [inline pkg/registry/apis/datasource/sub_health.go:65]: ^^ @marefr -- what is the right response here?

Comment (marefr) [reply pkg/registry/apis/datasource/sub_health.go:65]: 200 or 400 
https://github.com/grafana/grafana/blob/0b511aaace564461b1e117654eef1771c49cb3a7/pkg/api/datasources.go#L965-L969

**Review by wbrowne — COMMENTED** [review]
@ryantxu how come the separation of "config" and "connection"? As they both live under `apis/testdata.datasource.grafana.app/v0alpha1/namespaces/default`, I've missed where it's useful to have these split.

Also, tried this with the stack ID route - is there anything I'm missing to run that locally or do I need something more sophisticated to get me past `using a stack namespace requires deployment with a fixed stack id`?

Lastly, I assume no plans for the streaming side of things here - just punt on that for another day as no easy path to support this?

**Comment by ryantxu** [comment]
>  I've missed where it's useful to have these split.

With datasources, we have two very different use cases:
1. once configured, most people have access to see them and execute queries with them
2. datasource configuration and management.  We need to show the raw jsonData and some flavor of the secure stuff 

The kubernetes model for config is reasonably clear: it must look something like our current provisioning objects.  BUT this means that the full payload will be passed around everywhere (eg LIST).  This means anyone who can see the datasource will also see its config and at least some indication of the secure key part (still TBD, but likely ~what is shown in the edit UI now)

Connections?  These do not *need* to be modeled as k8s resources, but it seems convenient because it gives us a consistent user focused view of the same objects.

Splitting these has two values:
1. the view/usage case has a more streamlined shape -- does not include all the config stuff
2. we can have different authz patterns applied for connection vs config.  Specifically "viewer" should be able to see connections, but would not see config

**Comment by ryantxu** [comment]
> I assume no plans for the streaming side of things here - just punt on that for another day as no easy path to support this?

yup -- punting for now.  It is not required for the query execution stuff, so not urgent.  There are two main ways we could address it:
1. a global "live" service that streaming gets routed to -- this would map to our existing behavior with few changes.  Essentially a single websocket connection that can talk to anything
2. Manage "live" connections within each API server.  This is better because the explicit load/management is handled in the service, however it means clients would no longer have a single global websocket connection

but ya.. that we can handle another day :)

**Comment by ryantxu** [comment]
> tried this with the stack ID route - is there anything I'm missing to run that locally

You have to set it in custom.ini -- it is really just an alias, and will use orgId=1.  The benefit is that long term it is a namespace that can live in a multi-tenant server
```
[environment]
stack_id = abcd
```

image

**Comment by ryantxu** [comment]
@wbrowne -- re config vs connections, I think we should start with just connections and only add config when we have a real answer for secure data.

I have both in this PR because i think it is useful to keep in mind

**Comment by ryantxu** [comment]
Closing while we focus on https://github.com/grafana/grafana/pull/79726

**Pull request closed without merging**